### PR TITLE
Check feature flag on backend /graph PUT API for block permission

### PR DESCRIPTION
**Pull request by jeff303**
Check for the `:advanced-permissions` flag in API handler, when a `:schemas` `:block` entry is present in the graph

Set feature flag in tests when using API

Add new test to ensure failure in the case that feature flag is missing

### CODE REVIEW

File: enterprise/backend/test/metabase_enterprise/enhancements/models/permissions/block_permissions_test.clj [hunk 1]
```diff
@@ -139,7 +152,8 @@
                                       {:schemas :block, :native :write})]
           (is (schema= {:message  #".*DB permissions with a valid combination of values for :native and :schemas.*"
                         s/Keyword s/Any}
-                       (mt/user-http-request :crowberto :put 500 "permissions/graph" new-graph))))))))
+                       (premium-features-test/with-premium-features #{:advanced-permissions}
+                         (mt/user-http-request :crowberto :put 500 "permissions/graph" new-graph)))))))))
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/enhancements/models/permissions/block_permissions_test.clj:156]: This is not part of this PR but I think we probably should be returning a 400 instead of a 500 here

File: src/metabase/api/permissions.clj [hunk 2]
```diff
@@ -46,6 +48,16 @@
...
+
+    (let [has-block-perms? (atom false)]
+      (walk/postwalk (fn [x]
+                       (if (and (map? x) (= (:schemas x) :block))
+                         (do (reset! has-block-perms? true)
+                             nil)
+                         x)) graph)
+      (when (and @has-block-perms? (not (premium-features/has-feature? :advanced-permissions)))
+        (throw (ex-info (tru "Can''t use block permissions without having the advanced-permissions premium feature")
+                        {:status-code 403}))))
```
Comment (camsaul) [inline src/metabase/api/permissions.clj:60]: I've been using [`402 Payment Required`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402) for the other feature-flagged responses. Not 100% sure it makes sense but it does at least make the tests easier to write because we can confirm something is failing because of token feature flags rather than perms

Comment (camsaul) [reply src/metabase/api/permissions.clj:60]: In my impl for the fix I just added this check to `update-db-permissions!` in the `:block` code... that way you don't have to do any walking

**Review by camsaul — APPROVED** [review]
This fix works for me. I think you could make it a little simpler by doing the check inside `update-db-permissions!` when you see `:block` so you don't have to walk the whole thing.

It would also be nice to check for `:segmented` permissions and the `:sandboxes` feature -- but that's always worked that way so fixing it before the new release isn't as much of a priority

File: src/metabase/api/permissions.clj (hunk 2 above)
Comment (jeff303) [reply src/metabase/api/permissions.clj]: Fixed both

File: enterprise/backend/test/metabase_enterprise/enhancements/models/permissions/block_permissions_test.clj (hunk 1 above)
Comment (jeff303) [reply enterprise/backend/test/metabase_enterprise/enhancements/models/permissions/block_permissions_test.clj:156]: Yeah, let's tackle that in a separate PR.
